thread.c

```diff
@@ -581,6 +581,74 @@ static void thread_libevent_process(evutil_socket_t fd, short which, void *arg)
 /* Which thread we assigned a connection to most recently. */
 static int last_thread = -1;
 
+/* Last thread we assigned to a connection based on napi_id */
+static int last_thread_by_napi_id = -1;
+
+static LIBEVENT_THREAD *select_thread_round_robin(void)
+{
+    int tid = (last_thread + 1) % settings.num_threads;
+
+    last_thread = tid;
+
+    return threads + tid;
+}
+
+static void reset_threads_napi_id(void)
+{
+    LIBEVENT_THREAD *thread;
+    int i;
+
+    for (i = 0; i < settings.num_threads; i++) {
+         thread = threads + i;
+         thread->napi_id = 0;
+    }
+
+    last_thread_by_napi_id = -1;
+}
+
+/* Select a worker thread based on the NAPI ID of an incoming connection
+ * request. NAPI ID is a globally unique ID that identifies a NIC RX queue
+ * on which a flow is received.
+ */
+static LIBEVENT_THREAD *select_thread_by_napi_id(int sfd)
+{
+    LIBEVENT_THREAD *thread;
+    int napi_id, err, i;
+    socklen_t len;
+    int tid = -1;
+
+    len = sizeof(socklen_t);
+    err = getsockopt(sfd, SOL_SOCKET, SO_INCOMING_NAPI_ID, &napi_id, &len);
+    if ((err == -1) || (napi_id == 0))
+        return select_thread_round_robin();
```

Comment:
so, can you be stuck permanently in this case? I feel like I had a complaint on the original thread where you can enter a failure mode like this but there's no way for the user to know the feature isn't working.

Reply:
This can happen only if the user specifies -N option and the requests are coming from a virtual interface like loopback.

Reply:
so how does the user know what to do and how to fix it?

Reply:
It is a misconfiguration. The user should not use -N option if he is using virtual interfaces.  -N should be used only if the user is aware and can control the exact number of NIC receive queues through which the connections are received.
Should i add a counter to track this condition too?

Reply:
A counter or status code or something? I think I had this struggle with the original PR too: I think you folks are too close to the work. Imagine being some poor sod who downloads this thing and tries to make it work, except they don't work for intel and have never heard of or seen this before. Worse yet, memcached came from a package install so they'll have to be told to go out of their way to look up what a counter means or what an option means. Memcached has _millions_ of installs.

If a user starts a piece of software and it goes into a failure case like this, you get two outcomes:

1) User gets the same, or worse, performance as before. Gets mad at someone (me, you, etc) saying that the feature doesn't work.
2) User deploys without validating before/after performance. Doesn't have any idea if it's working or not, but it starts so they assume it does. Doesn't get mad at anyone but this is super depressing to imagine.

Or even worse still:
3) An admin deploys some change to the production fleet, which causes a driver reload/change/something and suddenly cluster performance tanks and their service/company goes down completely. There are no counters to compare and sparse documentation about what may have suddenly stopped working or why. This extends the amount of time it takes to get the website back online.

Which is a misery I've enjoyed at too many jobs. :)

Reply:
OK. I will add a counter and also add some documentation to the napi_ids.txt file to explain when we can hit this condition.